File: examples/python/procedures.py
"""Procedure hooks for the keyvalue grammar.

Each function fires after the corresponding variable is reduced.
Author-defined grammar hooks arrive namespaced as `hook_<name>` — the
grammar annotates Key with `@print`, and the entry point is `hook_print`,
so it can never collide with unrelated symbols.

These implementations are dispatched through the generated Python shim
(procedures_python.zig) by the extension module; no C is involved. The
build command `python3 <galley>/bindings/python/build.py <language-dir>`
detects this file (or `hooks/procedures.py`) and generates the shim
automatically, mirroring Go's `hooks/procedures.go` and Rust's
`procedures.rs`.
"""

import sys


def reduction(args):
    print("[hook] reduction", file=sys.stderr, flush=True)


def reduction_Document(args):
    print("[hook] Document", file=sys.stderr, flush=True)


def reduction_PairList(args):
    print("[hook] PairList", file=sys.stderr, flush=True)


def reduction_PairListTail(args):
    pass


def reduction_Pair(args):
    print("[hook] Pair", file=sys.stderr, flush=True)


def reduction_Key(args):
    print("[hook] Key", file=sys.stderr, flush=True)


def hook_print(args):
    print("[hook] print (Key)", file=sys.stderr, flush=True)


def reduction_KeyTail(args):
    pass


def reduction_Number(args):
    print("[hook] Number", file=sys.stderr, flush=True)


def reduction_NumberTail(args):
    pass

File: examples/typescript/procedures.ts
/**
 * Procedure hooks for the keyvalue grammar.
 *
 * Each function fires after the corresponding variable is reduced.
 * Author-defined grammar hooks arrive namespaced as `hook_<name>` — the
 * grammar annotates Key with `@print`, and the entry point is `hook_print`,
 * so it can never collide with unrelated symbols.
 *
 * These implementations are dispatched through the generated TypeScript shim
 * (procedures_typescript.zig) by the JS runtime; no C is involved. The
 * build command `npx galley-typescript-bindings <language-dir>` detects this
 * file and generates the shim automatically, mirroring Rust's
 * `procedures.rs`.
 */

import type { ProcedureArguments } from "galley-typescript-bindings";

function note(label: string, args: ProcedureArguments): void {
  const node = args.currentNode();
  if (node !== null) {
    node.text();
    node.children();
  }
  process.stderr.write(`[hook] ${label}\n`);
}

export function reduction(args: ProcedureArguments): void {
  note("reduction", args);
}

export function reduction_Document(args: ProcedureArguments): void {
  note("Document", args);
}

export function reduction_PairList(args: ProcedureArguments): void {
  note("PairList", args);
}

export function reduction_PairListTail(_args: ProcedureArguments): void {}

export function reduction_Pair(args: ProcedureArguments): void {
  note("Pair", args);
}

export function reduction_Key(args: ProcedureArguments): void {
  note("Key", args);
}

export function hook_print(args: ProcedureArguments): void {
  note("print (Key)", args);
}

export function reduction_KeyTail(_args: ProcedureArguments): void {}

export function reduction_Number(args: ProcedureArguments): void {
  note("Number", args);
}

export function reduction_NumberTail(_args: ProcedureArguments): void {}
